Don't block caret show events

feerrenrut opened the pull request:
Dont block caret show events.
When simplifying show/hide events, always leave the last state change in
the list of events.

### Link to issue number:
#511 "braille in Arabic does not display while writing"

### Summary of the issue:
When typing Arabic characters (using a keyboard and Arabic input) in wordpad (with left-align) the characters are not shown on braille display until space is pushed. 
This also affects typing English characters in wordpad (with right-align) .

### Description of how this pull request fixes the issue:
There are no "location changed" events when the characters are being typed, this makes sense since the characters are appearing to the right of the cursor (in the Arabic input case). I found that there are a series of "hide" then "show" events for the caret object when typing characters. I found a few places that these were being filtered out.

In `AddEvent` we were looking for a matching `hide` or `show` event and deleting the old event, however the new event was never added. I think it makes sense to keep the final event in a chain of `hide` and `show` events, so I removed the early exit from this condition. Without this part of the change the bug is not fixed.

### Testing performed:
Tested Arabic input in left and right aligned word pad documents.
Tested English input in left and right aligned word pad documents.

### Known issues with pull request:
I'm a little uncertain of the scope of behaviour that will be affected after making this change.

### Change log entry:
In Bug fixes:
> Braille displays are now updated when typing Arabic characters in a left-aligned WordPad document (#511).

feerrenrut committed:
Dont block caret show events

Fix for issue #511.
Dont block caret show events.
When simplifying show/hide events, always leave the last state change in
the list of events.

michaelDCurran reviewed (approved):
This certainly looks okay to me. Specifically you're just ensuring that Show events for the caret are not being dropped early.  I don't believe there would be any major side affects to this.

jcsteh commented:
A couple of thoughts:
1. I agree it makes sense to keep the show when we get (hide, show), since
this means it might have moved or at least blinked. That is really the crux
of the issue here. However, I don't think it makes sense to keep a hide
when there's a (show, hide), since the user could never have
seen/interacted with this; the hide just cancels out the show.
2. I notice you're bypassing shouldAcceptEvent for caret show events. Was
shouldAcceptEvent actually returning False here? If so, it'd be good to
know why. We do want to filter out events from background windows, etc. I
don't see why caret events should be an exception.

feerrenrut commented:
Keeping `hide` was for a few reasons:

- To provide symmetry with the `show` implementation.
- We can only say that the events cancel each other out if we are sure that they are unambiguously indicative of a state change. I don't think we're sure of this. For instance we may receive `show` events when the object is already in a "shown" state, or `hide` events when we are already in a "hidden" state. My investigation seemed to suggest that these events were commonly coming in pairs (several pairs) all at the same time stamp. These did not correspond with the caret's visible flashing.
- Granted, we don't (as far as I am aware) do anything with the caret `hide` events, we may want to in the future. By simplifying the events down to the final state change we are already simplifying (and optimising this a great deal). I counted 7 or so show/hide events in one event pump, reducing this to a single event is quite a reduction. If there is some chance we want to rely on `hide` events in the future, I suggest we don't block this here.

`shouldAcceptEvent` was returning `False`. We discard all `show`events unless the window class is one of the following:
```
	if eventName == "show":
		# Only accept 'show' events for specific cases, as otherwise we get flooded.
		return wClass in (
			"Frame Notification Bar", # notification bars
			"tooltips_class32", # tooltips
			"mscandui21.candidate", "mscandui40.candidate", "MSCandUIWindow_Candidate", # IMM candidates
			"TTrayAlert", # 5405: Skype
		)
```
See https://github.com/nvaccess/nvda/blob/master/source/eventHandler.py#L235

jcsteh commented:
[**@feerrenrut**](https://github.com/feerrenrut) commented on [4 Jul 2017, 15:14 GMT+10](https://github.com/nvaccess/nvda/pull/7332#issuecomment-312782083 "2017-07-04T05:14:05Z - Replied by Github Reply Comments"):
> For instance we may receive `show` events when the object is already in a "shown" state, or `hide` events when we are already in a "hidden" state.

I guess I'm arguing that a hide event when the object is already hidden is not going to be useful. The fact that the object is hidden means it's no longer a part of the UI, so NVDA shouldn't try to react to it in any way. In contrast, a show event when the object is already shown might suggest it moved and it certainly makes sense to react to that.

> My investigation seemed to suggest that these events were commonly coming in pairs (several pairs) all at the same time stamp. These did not correspond with the caret's visible flashing.

The fact that the caret "blinks" might be enough to suggest that hide is still possibly relevant for the caret.

At this stage, I'm inclined to agree: it might be extraneous, but trying to debug why we're missing a hide event in future could be really painful, and this alone is probably enough to justify this behaviour.

> `shouldAcceptEvent` was returning `False`. We discard all `show`events unless the window class is one of the following:

Oh blah. I should have figured that out, especially since I wrote that code. Sorry for the stupid question. :(

feerrenrut commented:
> Oh blah. I should have figured that out, especially since I wrote that code. Sorry for the stupid question. :(

Not a stupid question, its probably something I should have documented on the PR. It certainly helps to explain the need to bypass `shouldAcceptEvent`

feerrenrut committed:
Review actions #7332

Add comment to clarify why `shouldAcceptEvent` is bypassed.

leonardder commented:
This causes a regression in NP++ @BramD pointed me at. The braille cursor doesn't blink properly in NP++ any more.

bramd commented:
An additional observation in Notepad++:

It seems NVDA thinks there are cursor changes and so it almost immediately brings the braille display back to the cursor when I'm trying to scroll.

jcsteh commented:
I guess it fires show events whenever it blinks the caret, which we map to
caret events. However, we only expect caret events when the caret moves,
not when it blinks.

feerrenrut committed:
Fix braille scrolling in Notepad++ (#7149)

Fix only affects Notepad++. App modules can now override whether or not
we process caret show/hide events.

jcsteh commented on source/appModules/notepad++.py:
```diff
@@ -0,0 +1,27 @@
+class NotepadPlusPlusScintillaEdit(NVDAObject):
+
+	def initOverlayClass(self):
+		pass
```
This is a no-op and things will behave just fine if this method isn't present, so it can be removed.

jcsteh commented on source/appModules/notepad++.py:
```diff
@@ -0,0 +1,27 @@
+class AppModule(appModuleHandler.AppModule):
+
+	def event_NVDAObject_init(self, obj):
+		pass
```
Please remove this method.

jcsteh commented on source/appModules/notepad++.py:
```diff
@@ -0,0 +1,27 @@
+		pass
+
+	def chooseNVDAObjectOverlayClasses(self, obj, clsList):
+		clsList.insert(0, NotepadPlusPlusScintillaEdit)
```
This should only be inserted if the object is definitely the editor. Right now, this will be inserted for all objects; e.g. dialogs, etc. For example:

```
if obj.role == controlTypes.ROLE_EDITABLETEXT and obj.windowClassName == "Scintilla":
    clsList.insert(0, NotepadPlusPlusScintillaEdit)
```

jcsteh commented on source/IAccessibleHandler.py:
```diff
@@ -856,13 +856,18 @@ def pumpAll():
 		# later it will be mapped to a caret event, so skip `shouldAcceptEvent`
-		if not showCaretEvent and not eventHandler.shouldAcceptEvent(winEventIDsToNVDAEventNames[winEvent[0]], windowHandle=winEvent[1]):
+		if showHideCaretEvent:
+			if not focus.shouldAcceptShowHideCaretEvent():
```
I think this would be nicer as an attribute rather than a method. This way, classes can just set it to True or False as required if they don't need a condition. Those classes that do need a condition can implement it as a property. Note that we do the same for `shouldAllowIAccessibleFocusEvent`.

jcsteh commented on source/appModules/notepad++.py:
```diff
@@ -0,0 +1,27 @@
+		pass
+
+	def shouldAcceptShowHideCaretEvent(self):
+		return False
```
If this were an attribute as above, this could just be: `shouldAcceptShowHideCaretEvent = False`

jcsteh commented on source/NVDAObjects/__init__.py:
```diff
@@ -813,6 +813,13 @@ def _get_isCurrent(self):
 		"""
 		return False
 
+	def shouldAcceptShowHideCaretEvent(self):
```
Just a note that this feels weird being on the base NVDAObject rather than IAccessible, but right now, I understand that's how it has to be because we fire MSAA caret events on the focus object regardless of whether it's IAccessible. I think we should consider restricting these caret events to focus objects that are IAccessible subclasses in future, but that change is probably too risky for this PR. CC @MichaelDCurran for his thoughts.

tspivey commented:
This also breaks the review cursor in Notepad2 and the PureBasic IDE, both using the Scintilla edit control.
It jumps back to the position of the system cursor when moved.

feerrenrut committed:
Review actions for Notepad++ specific change

feerrenrut committed:
Apply change to all scintilla editors

Since there are several other editors that are suffering from the same
bug (Notepad2, PureBasicIDE). Apply the change to all scintilla editors.
This may mean that in some Scintilla based editors the initial problem
of braille not updating as characters are typed with left to right
languages in a right aligned document (or right to left languages in a
left aligned document) will persist. So far however, we are unaware of
any scintilla based editors that suffer from this problem.

feerrenrut commented:
@tspivey Thanks for pointing that out. We will apply this change to all scintilla edit controls. I chose to target Notepad++ specifically because I could reproduce the issue with SciTE. This may mean that the initial problem (braille not updating as characters are typed when typing a left to right language in a right aligned document, or right to left languages in a left aligned document) may persist for some scintilla based editor. It should be noted however, at this point we have not found a Scintilla based editor that is affected by this problem.

feerrenrut commented:
@tspivey I have pushed changes that should resolve, this would you mind testing this again with Notepad2 and PureBasic IDE?

tspivey commented:
I assume you meant couldn't reproduce the issue with SciTE -- I also tried with that and couldn't reproduce the issue.

Unfortunately, running directly from origin/i511-brailleInArabic, this doesn't fix Notepad2 or PureBasic.

jcsteh commented:
@feerrenrut, I approved the change here from a code perspective, but I missed the second paragraph of @tspivey's [comment](https://github.com/nvaccess/nvda/pull/7332#issuecomment-321221412), which might obviously impact whether this is ready. :)

feerrenrut committed:
Commit missing hunk.

Forgot to add a hunk when commiting. Which totally negated the intended
fix.

feerrenrut commented:
Oops my mistake. I forgot to add the part of this which was required to actually ignore these events for Scintilla based edit controls. I have also tested this with Notepad 2. @tspivey Feel free to test this branch again.

jcsteh reviewed (approved):
Wow. I ... totally missed that when I reviewed. This is definitely the more correct change. :)
